translate/storage/lisa.py

```diff
@@ -329,20 +333,40 @@ def serialize_hook(treestring):
     def serialize(self, out=None):
         """Converts to a string containing the file's XML"""
         root = self.document.getroot()
+        xml_quote_format = "'"
+        xml_encoding = self.encoding.lower()
+
         if self.XMLdoublequotes:
-            if self.XMLuppercaseEncoding:
-                out.write(b'<?xml version="1.0" encoding="UTF-8"?>\n')
-            else:
-                out.write(b'<?xml version="1.0" encoding="utf-8"?>\n')
+            xml_quote_format = '"'
+
+        if self.XMLuppercaseEncoding:
+            xml_encoding = self.encoding.upper()
+
+        xml_declaration = f"<?xml version={xml_quote_format}1.0{xml_quote_format} encoding={xml_quote_format}{xml_encoding}{xml_quote_format}?>\n"
+
+        out.write(xml_declaration.encode(self.encoding))
+
         if self.XMLindent:
             reindent(root, **self.XMLindent)
-        treestring = etree.tostring(
-            self.document,
-            pretty_print=not self.XMLindent,
-            xml_declaration=not self.XMLdoublequotes,
-            encoding=self.encoding,
-            doctype=self.XMLdoctype,
-        )
+        # When XMLSelfClosingTags flag is active, relay
+        # against LXML library - otherwise we're using
+        # standard xml library.
+        if self.XMLSelfClosingTags:
+            treestring = etree.tostring(
+                self.document,
+                pretty_print=not self.XMLindent,
+                encoding=self.encoding,
+                doctype=self.XMLdoctype,
+            )
+        else:
+            if self.XMLdoctype:
+                out.write(b"" + self.XMLdoctype + "\n")
+            ElementTree.register_namespace("", self.namespace)
```

Comment:
This worries me as the registry is global and can influence other ElementTree uses. It is nearly impossible to test that this doesn't break anything inside complex applications like Weblate.

Maybe it would be better to use `default_namespace` of `tostring`? Unfortunately, it is Python 3.8+ only.

See https://docs.python.org/3/library/xml.etree.elementtree.html#xml.etree.ElementTree.register_namespace and https://docs.python.org/3/library/xml.etree.elementtree.html#xml.etree.ElementTree.tostring

Reply:
My python knowledge does not allow me to make such arbitrary 😶

both translate and Weblate are requiring python 3.7 as a minimum version which is mostly the reason why I didn't use this way.